### PR TITLE
feat(ui): add passcode input component

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Add passcode input component

image

image

This component Include following feature:
- autofocus to the next input field after code input
- auto-delete the previous input field value when keep clicking on the Backspace key
- numeric input only
- navigate throw input field using ArrowLeft and ArrowRight key
- auto paste the input fields from the clipboard
- auto-update current field value when typing new chars

## Linear Issue Reference

LOG-1766

## Testing

UT cases added
test locally
@logto-io/eng

### CODE REVIEW

File: packages/ui/src/components/Passcode/index.tsx [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+  return newValue;
+};
+
+const Passcode = ({
+  name,
+  isDisabled,
+  className,
+  value,
+  length = 6,
```
Comment (demonzoo) [inline packages/ui/src/components/Passcode/index.tsx:54]: `length = defaultLength`?

File: packages/ui/src/components/Passcode/index.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+      // Update the total input value
+      onChange(Object.assign([], codes, { [targetId]: trim(codes[targetId], value) }));
+
+      // Move to the next target
+      if (value) {
+        const nextTarget = inputReferences.current[targetId + 1];
+        nextTarget?.focus();
+        nextTarget?.select();
+      }
```
Comment (demonzoo) [inline packages/ui/src/components/Passcode/index.tsx:98]: what if it's already the last input? Will it be index out of range?

File: packages/ui/src/components/Passcode/index.tsx (hunk 1 above)
Comment (simeng-li) [reply packages/ui/src/components/Passcode/index.tsx]: thx for catching that

File: packages/ui/src/components/Passcode/index.tsx [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  const inputs = useMemo(
+    () =>
+      Array.from({ length }).map((_, index) => (
+        <input
+          ref={(element) => {
+            // eslint-disable-next-line @silverhand/fp/no-mutation
+            inputReferences.current[index] = element;
+          }}
+          // eslint-disable-next-line react/no-array-index-key
+          key={`${name}_${index}`}
```
Comment (demonzoo) [inline packages/ui/src/components/Passcode/index.tsx:198]: Where does `name` come from? Didn't see the declaration in this `useMemo` block

File: packages/ui/src/components/Passcode/index.tsx (hunk 2 above)
Comment (simeng-li) [reply packages/ui/src/components/Passcode/index.tsx:98]: yep, considering handling it on the upper scope.  Gonna have a container component wrap up the UI, will handle the code format validation, error display, count down, resend and submit logic. 
Might have a callback method props added later or use an onChange useEffect hook to handle that.  
 i.e. if the user updates a digit in the middle, should we `focus on the submit button ` or `auto submit` as well.

**Review by wangsijie — APPROVED** [review]
Nice work 🍺

File: packages/ui/src/components/Passcode/index.tsx (hunk 2 above)
Comment (demonzoo) [reply packages/ui/src/components/Passcode/index.tsx:98]: ok

**Pull request merged**